[OpenMP] Use half of available logical processors for collapse tests

xingxue-ibm opened the pull request:
The new collapse test cases define `MAX_THREADS` to be 256 and use all available threads/logical processors on the system. This triples the testing time on an AIX machine that has 128 logical processors. This patch changes to use half of available logical processors to avoid over subscribing because there are other libomp tests running at the same time, including 2 other such collapse tests.
```
Slowest Tests:
--------------------------------------------------------------------------
971.92s: libomp :: worksharing/for/omp_collapse_many_LTLEGE_int.c
971.62s: libomp :: worksharing/for/omp_collapse_many_GTGEGT_int.c
965.29s: libomp :: worksharing/for/omp_collapse_many_GELTGT_int.c
332.97s: libomp :: tasking/omp_taskloop_num_tasks.c
233.26s: libomp :: worksharing/single/omp_single.c
207.42s: libomp :: tasking/omp_taskloop_grainsize.c
178.41s: libomp :: tasking/hidden_helper_task/depend.cpp
172.02s: libomp :: worksharing/for/omp_collapse_many_int.c
....
```

xingxue-ibm committed:
Use lower max threads to reduce the testing time on AIX.

jprotze commented:
The test should not use more than 128 threads on your system (default value returned by `omp_get_max_threads` is number of available cores). Since we usually run multiple test in parallel, I think it makes sense to modify the test to never use more than half of the available cores:

```diff
-  unsigned num_threads = omp_get_max_threads();
+  unsigned num_threads = omp_get_max_threads()/2;
  if (num_threads > MAX_THREADS)
    num_threads = MAX_THREADS;
  omp_set_num_threads(num_threads);
```

Furthermore, I suggest to move the `omp_get_thread_num` call out of the loop:

```diff
#pragma omp parallel num_threads(num_threads)
  {
+    unsigned gtid = omp_get_thread_num();
#pragma omp for collapse(3) private(i, j, k)
    LOOP {
      unsigned count;
-      unsigned gtid = omp_get_thread_num();
```

xingxue-ibm commented:
> The test should not use more than 128 threads on your system (default value returned by `omp_get_max_threads` is number of available cores). Since we usually run multiple test in parallel, I think it makes sense to modify the test to never use more than half of the available cores:
> 
> ```diff
> -  unsigned num_threads = omp_get_max_threads();
> +  unsigned num_threads = omp_get_max_threads()/2;
>   if (num_threads > MAX_THREADS)
>     num_threads = MAX_THREADS;
>   omp_set_num_threads(num_threads);
> ```
> 
> Furthermore, I suggest to move the `omp_get_thread_num` call out of the loop:
> 
> ```diff
> #pragma omp parallel num_threads(num_threads)
>   {
> +    unsigned gtid = omp_get_thread_num();
> #pragma omp for collapse(3) private(i, j, k)
>     LOOP {
>       unsigned count;
> -      unsigned gtid = omp_get_thread_num();
> ```
Good point, thanks @jprotze! It makes sense not to use more than half of the available logical processors to avoid over-subscribing, noting there are other libomp test cases run running in parallel, including 2 other collapse test cases using `collapse_test.inc`. Changed as per suggestion. Also removed `omp_set_num_threads(num_threads);` because it will make the call to `omp_get_max_threads()` in the next iteration return half of the original value.

jprotze commented:
Did you push the changes?

xingxue-ibm commented:
> Did you push the changes?

Having a problem pushing the changes.  Working on it...

xingxue-ibm committed:
Addressed comments.
* Use half of available threads/logical processors to avoid over-subscribing
* Fix the PRINTF macro to get rid of warnings

xingxue-ibm commented:
> > Did you push the changes?
> 
> Having a problem pushing the changes. Working on it...

Changes are pushed.

jprotze commented on openmp/runtime/test/worksharing/for/collapse_test.inc:
```diff
@@ -106,12 +106,14 @@ int test() {
     num_threads = MAX_THREADS;
-  omp_set_num_threads(num_threads);
+
+  unsigned *chunkSizesOpenmp = malloc(sizeof(unsigned) * num_threads);
```
If you malloc, you should also free. I don't see anything wrong with the previous stack allocation approach.

To reduce the impact of cache line sharing, it would help to use long instead of int.

xingxue-ibm replied:
> If you malloc, you should also free. I don't see anything wrong with the previous stack allocation approach.
> 
Good catch, will fix it. Thanks!
> To reduce the impact of cache line sharing, it would help to use long instead of int.

Did you mean to change `unsigned *chunkSizesOpenmp` to `unsigned long *chunkSizeOpenmp`?

xingxue-ibm replied:
>>  To reduce the impact of cache line sharing, it would help to use long instead of int.
>
> Did you mean to change unsigned *chunkSizesOpenmp to unsigned long *chunkSizeOpenmp?

Hi @jprotze, I've addressed comments except the one quoted above. It would be appreciated if you can clarify at your convenience.

jprotze replied:
Yes, that's the change I proposed.

xingxue-ibm replied:
Thanks, @jprotze! I've changed the type of `chunkSizeOpenmp` to `unsigned long` and refreshed the patch.

jpeyton52 commented on openmp/runtime/test/worksharing/for/collapse_test.inc:
```diff
@@ -106,12 +106,14 @@ int test() {
-  unsigned chunkSizesOpenmp[MAX_THREADS] = {0};
 
-  unsigned num_threads = omp_get_max_threads();
+  // Use half of the available threads/logical processor.
+  unsigned num_threads = omp_get_max_threads() / 2;
   if (num_threads > MAX_THREADS)
     num_threads = MAX_THREADS;
```
Because the test reduces `num_threads`, please also add a check ensuring num_threads is not 0. Machines with only one hardware thread (or only one logical processor assigned to the process) will now assign num_threads to 0.

```
if (num_threads == 0)
  num_threads = 1;
```

xingxue-ibm replied:
> Because the test reduces `num_threads`, please also add a check ensuring num_threads is not 0. Machines with only one hardware thread (or only one logical processor assigned to the process) will now assign num_threads to 0.
> 
> ```
> if (num_threads == 0)
>   num_threads = 1;
> ```

Good catch, added the check. Thanks!

xingxue-ibm committed:
Addressed comments.
* add free()
* make sure num_threads is not 0

xingxue-ibm committed:
Merge branch 'llvm:main' into main

xingxue-ibm committed:
Merge branch 'llvm:main' into main

xingxue-ibm committed:
Addressed comments.
- make chunkSizesOpenmp type 'unsinged long'.

jprotze reviewed (approved):
Lgtm now
